src/internal/ans.ts

```diff
@@ -147,29 +149,359 @@ export async function registerName(args: RegisterNameParameters): Promise<InputS
+      functionArguments: [AccountAddress.fromRelaxed(address).toString()],
+    },
+  });
+
+  const domainName = unwrapOption<MoveAddressType>(res[1]);
+  const subdomainName = unwrapOption<MoveAddressType>(res[0]);
+
+  if (!domainName) return undefined;
+
+  return [subdomainName, domainName].filter(Boolean).join(".");
+}
+
+export async function setPrimaryName(args: {
+  aptosConfig: AptosConfig;
+  sender: Account;
+  name: string | null;
+  options?: InputGenerateTransactionOptions;
+}): Promise<InputSingleSignerTransaction> {
+  const { aptosConfig, sender, name, options } = args;
+  const routerAddress = getRouterAddress(aptosConfig);
+
+  if (!name) {
+    const transaction = await generateTransaction({
+      aptosConfig,
+      sender: sender.accountAddress.toString(),
+      data: {
+        function: `${routerAddress}::router::clear_primary_name`,
+        functionArguments: [],
+      },
+      options,
+    });
+
+    return transaction as InputSingleSignerTransaction;
+  }
+
+  const { domainName, subdomainName } = isValidANSName(name);
+
+  const transaction = await generateTransaction({
+    aptosConfig,
+    sender: sender.accountAddress.toString(),
+    data: {
+      function: `${routerAddress}::router::set_primary_name`,
+      functionArguments: [domainName, subdomainName],
+    },
+    options,
+  });
+
+  return transaction as InputSingleSignerTransaction;
+}
+
+export async function getTargetAddress(args: {
+  aptosConfig: AptosConfig;
+  name: string;
+}): Promise<MoveAddressType | undefined> {
+  const { aptosConfig, name } = args;
+  const routerAddress = getRouterAddress(aptosConfig);
+  const { domainName, subdomainName } = isValidANSName(name);
+
+  const res = await view({
+    aptosConfig,
+    payload: {
+      function: `${routerAddress}::router::get_target_addr`,
+      functionArguments: [domainName, Option(subdomainName)],
+    },
+  });
+
+  const target = unwrapOption<MoveAddressType>(res[0]);
+  return target ? AccountAddress.fromRelaxed(target).toString() : undefined;
+}
+
+export async function setTargetAddress(args: {
+  aptosConfig: AptosConfig;
+  sender: Account;
+  name: string;
+  address: AccountAddressInput;
+  options?: InputGenerateTransactionOptions;
+}): Promise<InputSingleSignerTransaction> {
+  const { aptosConfig, sender, name, address, options } = args;
+  const routerAddress = getRouterAddress(aptosConfig);
+  const { domainName, subdomainName } = isValidANSName(name);
+
+  const transaction = await generateTransaction({
+    aptosConfig,
+    sender: sender.accountAddress.toString(),
+    data: {
+      function: `${routerAddress}::router::set_target_addr`,
+      functionArguments: [domainName, subdomainName, address],
+    },
+    options,
+  });
+
+  return transaction as InputSingleSignerTransaction;
+}
+
+export async function getName(args: {
+  aptosConfig: AptosConfig;
+  name: string;
+}): Promise<GetANSNameResponse[0] | undefined> {
```

Comment:
dont think you need to specify `undefined` type, can returns the first index in the array `[0]` and if ti is empty it would return `undefined`

Reply:
These are only compile time types, it won't have access to check if the array is empty or not at runtime.